src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionEvalStarlarkThreadContext.java

```diff
@@ -56,94 +59,164 @@ public static ModuleExtensionEvalStarlarkThreadContext fromOrNull(StarlarkThread
     this.basePackageId = basePackageId;
-    this.repoMapping = repoMapping;
+    this.baseRepoMapping = baseRepoMapping;
     this.directories = directories;
     this.eventHandler = eventHandler;
   }
 
-  public void createRepo(StarlarkThread thread, Dict<String, Object> kwargs, RuleClass ruleClass)
-      throws InterruptedException, EvalException {
+  /**
+   * Records a call to a repo rule that should be created at the end of the module extension
+   * evaluation.
+   */
+  public void lazilyCreateRepo(
+      StarlarkThread thread, Dict<String, Object> kwargs, RuleClass ruleClass)
+      throws EvalException {
     Object nameValue = kwargs.getOrDefault("name", Starlark.NONE);
     if (!(nameValue instanceof String name)) {
       throw Starlark.errorf(
           "expected string for attribute 'name', got '%s'", Starlark.type(nameValue));
     }
     RepositoryName.validateUserProvidedRepoName(name);
-    RepoSpecAndLocation conflict = generatedRepos.get(name);
+    RepoRuleCall conflict = deferredRepos.get(name);
     if (conflict != null) {
       throw Starlark.errorf(
           "A repo named %s is already generated by this module extension at %s",
-          name, conflict.getLocation());
+          name, conflict.location());
     }
-    String prefixedName = repoPrefix + name;
-    try {
-      Rule rule =
-          BzlmodRepoRuleCreator.createRule(
-              basePackageId,
-              repoMapping,
-              directories,
-              thread.getSemantics(),
-              eventHandler,
-              thread.getCallStack(),
-              ruleClass,
-              Maps.transformEntries(kwargs, (k, v) -> k.equals("name") ? prefixedName : v));
+    deferredRepos.put(
+        name,
+        new RepoRuleCall(
+            ruleClass,
+            // The extension may mutate the values of the kwargs after this function returns.
+            (Dict<String, Object>) deepCloneStarlarkObject(kwargs),
+            thread.getCallerLocation(),
+            thread.getCallStack()));
+  }
 
-      Map<String, Object> attributes =
-          Maps.filterKeys(
-              Maps.transformEntries(kwargs, (k, v) -> rule.getAttr(k)), k -> !k.equals("name"));
-      String bzlFile = ruleClass.getRuleDefinitionEnvironmentLabel().getUnambiguousCanonicalForm();
-      var attributesValue = AttributeValues.create(attributes);
-      AttributeValues.validateAttrs(
-          attributesValue, String.format("%s '%s'", rule.getRuleClass(), name));
-      RepoSpec repoSpec =
-          RepoSpec.builder()
-              .setBzlFile(bzlFile)
-              .setRuleClassName(ruleClass.getName())
-              .setAttributes(attributesValue)
-              .build();
+  /**
+   * Evaluates the repo rule calls recorded by {@link #lazilyCreateRepo} and returns all repos
+   * generated by the extension. The key is the "internal name" (as specified by the extension) of
+   * the repo, and the value is the package containing (only) the repo rule.
```

Comment:
the value is the RepoSpec these days. we should've updated the documentation before